File: src/main/java/no/skatteetaten/aurora/databasehotel/service/Janitor.java
package no.skatteetaten.aurora.databasehotel.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

@Component
public class Janitor {

    private static Logger LOG = LoggerFactory.getLogger(Janitor.class);

    private DatabaseHotelAdminService databaseHotelAdminService;

    public Janitor(DatabaseHotelAdminService databaseHotelAdminService) {
        this.databaseHotelAdminService = databaseHotelAdminService;
    }

    @Scheduled(fixedDelay = 3600 * 1000, initialDelay = 60 * 1000)
    public void deleteOldUnusedSchemas() {

        LOG.info("Periodic deletion of old unused schemas");
        databaseHotelAdminService.findAllDatabaseInstances().forEach(DatabaseInstance::deleteUnusedSchemas);
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/service/Janitor.kt
package no.skatteetaten.aurora.databasehotel.service

import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

private val logger = KotlinLogging.logger {}

@Component
class Janitor(
    private val databaseHotelAdminService: DatabaseHotelAdminService,
    @param:Value("\${database-config.dropAllowed}") private val dropAllowed: Boolean
) {

    @Scheduled(fixedDelay = 3600L * 1000, initialDelay = 60L * 1000)
    fun deleteStaleSchemasByCooldown() {

        if (!dropAllowed) return

        logger.info("Periodic run of janitor")
        databaseHotelAdminService.findAllDatabaseInstances().forEach {
            it.deactivateStaleSchemas()
            it.deleteSchemasWithExpiredCooldowns()
        }
    }
}
